packages/flutter/lib/src/cupertino/sliding_segmented_control.dart

```diff
@@ -0,0 +1,1037 @@
+  ///
+  /// Its value must be one of the keys in the [Map] of [children], or null, in
+  /// which case no widget will be selected.
+  ///
+  /// The [controller]'s value changes when the user drags the thumb to a different
+  /// child widget, or tap on a different child widget. When it is changed programmatically,
+  /// all sliding animations will play as if the new selected child widget was
+  /// tapped on.
+  final ValueNotifier<T> controller;
+
+  /// The color used to paint the rounded rect behind the [children] and the separators.
+  ///
+  /// The default value is [CupertinoColors.tertiarySystemFill]. Skips painting
+  /// entirely if null is specified.
+  final Color backgroundColor;
+
+  /// The color used to paint the interior of the thumb that appears behind the
+  /// currently selected item.
+  ///
+  /// The default value is a [CupertinoDynamicColor] that appears white in light
+  /// mode and becomes a gray color in dark mode.
+  final Color thumbColor;
+
+  /// The amount of space by which to inset the [children].
+  ///
+  /// Must not be null. Defaults to EdgeInsets.symmetric(vertical: 2, horizontal: 3).
+  final EdgeInsetsGeometry padding;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<CupertinoSlidingSegmentedControl<T>>
+    with TickerProviderStateMixin<CupertinoSlidingSegmentedControl<T>> {
+
+  final Map<T, AnimationController> _highlightControllers = <T, AnimationController>{};
+  final Tween<FontWeight> highlightTween = _FontWeightTween(begin: FontWeight.normal, end: FontWeight.w600);
+
+  final Map<T, AnimationController> _pressControllers = <T, AnimationController>{};
+  final Tween<double> pressTween = Tween<double>(begin: 1, end: 0.2);
+
+  TextDirection textDirection;
+
+  AnimationController thumbController;
+  AnimationController separatorOpacityController;
+  AnimationController thumbScaleController;
+
+  final HorizontalDragGestureRecognizer drag = HorizontalDragGestureRecognizer();
+
+  ValueNotifier<T> get controller => _controller;
+  ValueNotifier<T> _controller;
+  set controller(ValueNotifier<T> newController) {
+    assert(newController != null);
+    if (controller == newController)
+      return;
+    controller?.removeListener(_didChangeControllerValue);
+    newController.addListener(_didChangeControllerValue);
+  }
+
+  AnimationController createHighlightAnimationController({ bool isCompleted = false }) {
+    return AnimationController(
+      duration: _kHighlightAnimationDuration,
+      value: isCompleted ? 1 : 0,
+      vsync: this,
+    )..addListener(() {
+      setState(() {
+        // The textstyle has changed.
+      });
+    });
+  }
+
+  AnimationController createFadeoutAnimationController() {
+    return AnimationController(
+      duration: _kOpacityAnimationDuration,
+      vsync: this,
+    )..addListener(() {
+      setState(() {
```

Comment:
The number of `AnimationController`s is proportional to the number of segments. I thought each `AnimatedBuilder` can only be bound to one `Animation`?

Reply:
True, I'm not aware of a way to do that with AnimatedBuilder.

Reply:
But the list of child `Listenable`s seems to be immutable once the `_MergingListenable` is constructed using `Listenable.merge`. In my case I need to add new `Listenable`s and remove existing `Listenable`s from the list in `didUpdateWidget`, because developers may want to add/remove segments.

Reply:
If you build the Listenable with Listenable.merge() when you build the AnimatedBuilder itself, there's no getting out of sync?